docs/proposals/20230612-node-os-image-caching.md

```diff
@@ -116,74 +118,78 @@ Example of the enviornment variable being turned on:
 
 Diagram of the Node OS Caching Process:
 
 ![Figure 1](./images/node-os-image-cache.png)
 
-Example AzureMachinePool yaml with the timestamps in status:
+As for why the healthy node has to be shut down while creating a snapshot of it, if it isn’t shut down first then pods can be scheduled as the snapshot is taken which will cause some dangerous states in terms of how it exists after being utilized by the AzureMachineTemplates.
+
+In terms of how a healthy node would be selected, there is already state data present on each AzureMachinePoolMachine and AzureMachine which we can use and is listed in the examples below. An ideal node would be one which is running and healthy. Whichever node has been running and healthy for the longest amount of time since the last patch and has the patch applied should be chosen as it’s the most overall stable. This means that for AzureMachinePoolMachines and AzureMachines we will take the node with the earliest creation time from metadata.creationTimestamp and has latestModelApplied : true present if it is an AzureMachinePoolMachine. As the prototype is always from a successfully healthy and working node the image is always known to be working before being chosen for replication. In terms of knowing if the patch has been applied, for the first development of this feature we won't know but all nodes should roughly be patched together so it will be fine in most cases.
+
+Example AzureMachinePoolMachine yaml with the important fields shown (want status.ready == true, latestModelApplied: true and each status: "True"):
 
 ```yaml
 apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
 kind: AzureMachinePool
 metadata:
-  name: node-os-image-caching-machine-pool
+  name: node-os-image-caching-machine-pool-machine
   namespace: default
+  creationTimestamp: "2023-06-20T17:41:54Z"
 status:
   conditions:
-  - lastTransitionTime: "2023-06-12T23:14:55Z"
+  - lastTransitionTime: "2023-06-20T17:43:39Z"
     status: "True"
     type: Ready
-  - lastTransitionTime: "2023-06-12T23:14:00Z"
+  - lastTransitionTime: "2023-06-20T17:43:39Z"
     status: "True"
     type: BootstrapSucceeded
-  - lastTransitionTime: "2023-06-12T23:14:55Z"
-    status: "True"
-    type: ScaleSetDesiredReplicas
-  - lastTransitionTime: "2023-06-12T23:14:55Z"
-    status: "True"
-    type: ScaleSetModelUpdated
-  - lastTransitionTime: "2023-06-12T23:14:55Z"
+  - lastTransitionTime: "2023-06-20T17:43:39Z"
     status: "True"
-    type: ScaleSetRunning
+    type: NodeHealthy
+  latestModelApplied: true
+  ready: true
 ```
 
-Example AzureMachine yaml with the proposed additional needed timestamps in status:
+Example AzureMachine yaml with the important fields shown (want status.ready == true and each status: "True"):
```

Comment:
If an AzureMachine doesn't have something like `status.latestModelApplied`, then how do we know we're selecting a node running the most up-to-date image? Otherwise it seems like there's a risk that a node running an older image is selected for replication and future nodes are based on that older image.

Reply:
Added a line explaining this here: 'If it is an AzureMachine, since they tend to be updated together with automatic security updates and are immutable in regard to user changes to the OS image, we don't need to worry about a field like latestModelApplied.' I hope that kind of makes sense, basically latestModelApplied lets us know when the user has altered the image manually and it is still updating to that image, but AzureMachines don't behave like that in general, so we don't need to account for this 1:1 (at least as far as I currently understand it), we should just need typical ready conditions and creation timestamps.